File: frontend/apps/chinese/src/features/lessons/lib/quizScore.test.ts
import { describe, expect, it } from 'vitest'
import { buildAnswers, countUnanswered, firstUnansweredIndex, isPassed, minCorrectToPass, scorePercent } from './quizScore'
import type { QuizQuestion } from '@af/chinese-kit'

const q = (id: string): QuizQuestion => ({ id, type: 'single_choice', prompt: 'p', promptLang: 'vi', options: [] })

describe('scorePercent', () => {
  it('làm tròn XUỐNG (7/9 = 77, không phải 78)', () => {
    expect(scorePercent(7, 9)).toBe(77)
    expect(scorePercent(4, 5)).toBe(80)
    expect(scorePercent(6, 7)).toBe(85)
    expect(scorePercent(0, 7)).toBe(0)
    expect(scorePercent(7, 7)).toBe(100)
  })

  it('total = 0 hoặc correct âm ⇒ 0; correct > total kẹp về 100', () => {
    expect(scorePercent(3, 0)).toBe(0)
    expect(scorePercent(-1, 5)).toBe(0)
    expect(scorePercent(9, 5)).toBe(100)
  })
})

describe('isPassed', () => {
  it('so sánh số nguyên không làm tròn: 4/5 đạt, 7/9 không đạt', () => {
    expect(isPassed(4, 5)).toBe(true)
    expect(isPassed(7, 9)).toBe(false)
    expect(isPassed(8, 10)).toBe(true)
    expect(isPassed(5, 7)).toBe(false) // 71%
    expect(isPassed(6, 7)).toBe(true) // 85%
  })

  it('total = 0 ⇒ không đạt; ngưỡng tuỳ chỉnh', () => {
    expect(isPassed(0, 0)).toBe(false)
    expect(isPassed(1, 2, 50)).toBe(true)
  })
})

describe('minCorrectToPass', () => {
  it('cần đúng ≥ N câu', () => {
    expect(minCorrectToPass(5)).toBe(4)
    expect(minCorrectToPass(7)).toBe(6)
    expect(minCorrectToPass(9)).toBe(8)
    expect(minCorrectToPass(10)).toBe(8)
    expect(minCorrectToPass(0)).toBe(0)
  })
})

describe('countUnanswered / firstUnansweredIndex / buildAnswers', () => {
  const questions = [q('a'), q('b'), q('c')]

  it('đếm và tìm câu chưa trả lời', () => {
    const answers = new Map([['a', 'x'], ['c', 'y']])
    expect(countUnanswered(questions, answers)).toBe(1)
    expect(firstUnansweredIndex(questions, answers)).toBe(1)
    expect(firstUnansweredIndex(questions, new Map([['a', '1'], ['b', '2'], ['c', '3']]))).toBe(-1)
  })

  it('đóng gói theo thứ tự câu hỏi, bỏ câu chưa trả lời, không có đáp án thừa', () => {
    const answers = new Map([['c', 'z'], ['a', 'x'], ['zz', 'thừa']])
    expect(buildAnswers(questions, answers)).toEqual([
      { questionId: 'a', optionId: 'x' },
      { questionId: 'c', optionId: 'z' },
    ])
  })
})

File: mobile/apps/chinese/test/features/lessons/quiz_score_test.dart
import 'package:af_chinese/features/lessons/data/models.dart';
import 'package:af_chinese/features/lessons/domain/quiz_score.dart';
import 'package:flutter_test/flutter_test.dart';

QuizQuestion q(String id) => QuizQuestion(id: id, prompt: 'p');

// Chép đủ ca của `quizScore.test.ts` web (7 ca).
void main() {
  group('scorePercent', () {
    test('làm tròn XUỐNG (7/9 = 77, không phải 78)', () {
      expect(scorePercent(7, 9), 77);
      expect(scorePercent(4, 5), 80);
      expect(scorePercent(6, 7), 85);
      expect(scorePercent(0, 7), 0);
      expect(scorePercent(7, 7), 100);
    });

    test('total = 0 hoặc correct âm ⇒ 0; correct > total kẹp về 100', () {
      expect(scorePercent(3, 0), 0);
      expect(scorePercent(-1, 5), 0);
      expect(scorePercent(9, 5), 100);
    });
  });

  group('isPassed', () {
    test('so sánh số nguyên không làm tròn: 4/5 đạt, 7/9 không đạt', () {
      expect(isPassed(4, 5), isTrue);
      expect(isPassed(7, 9), isFalse);
      expect(isPassed(8, 10), isTrue);
      expect(isPassed(5, 7), isFalse); // 71%
      expect(isPassed(6, 7), isTrue); // 85%
    });

    test('total = 0 ⇒ không đạt; ngưỡng tuỳ chỉnh', () {
      expect(isPassed(0, 0), isFalse);
      expect(isPassed(1, 2, 50), isTrue);
    });
  });

  group('minCorrectToPass', () {
    test('cần đúng ≥ N câu', () {
      expect(minCorrectToPass(5), 4);
      expect(minCorrectToPass(7), 6);
      expect(minCorrectToPass(9), 8);
      expect(minCorrectToPass(10), 8);
      expect(minCorrectToPass(0), 0);
    });
  });

  group('countUnanswered / firstUnansweredIndex / buildAnswers', () {
    final questions = [q('a'), q('b'), q('c')];

    test('đếm và tìm câu chưa trả lời', () {
      final answers = {'a': 'x', 'c': 'y'};
      expect(countUnanswered(questions, answers), 1);
      expect(firstUnansweredIndex(questions, answers), 1);
      expect(firstUnansweredIndex(questions, {'a': '1', 'b': '2', 'c': '3'}), -1);
    });

    test('đóng gói theo thứ tự câu hỏi, bỏ câu chưa trả lời, không có đáp án thừa', () {
      final answers = {'c': 'z', 'a': 'x', 'zz': 'thừa'};
      expect(buildAnswers(questions, answers), [
        const QuizAnswer(questionId: 'a', optionId: 'x'),
        const QuizAnswer(questionId: 'c', optionId: 'z'),
      ]);
    });
  });
}
